src/validation/validation.cpp

```diff
@@ -2508,6 +2511,17 @@ bool ConnectBlockCanonicalOrdering(const CBlock &block,
                         REJECT_INVALID, "bad-txns-nonfinal");
                 }
 
+                if (fStrictPayToScriptHash)
+                {
+                    // Add in sigops done by pay-to-script-hash inputs;
+                    // this is to prevent a "rogue miner" from creating
+                    // an incredibly-expensive-to-validate block.
+                    nSigOps += GetP2SHSigOpCount(txref, view, flags);
+                    if (nSigOps > GetMaxBlockSigOpsCount(block.GetBlockSize()))
```

Comment:
let's combine this check and the one above by putting them both outside the "if (!tx.IsCoinBase())" block (and for DTOR as well).